src/librbd/api/Mirror.cc

```diff
@@ -2098,6 +2345,1689 @@ void Mirror<I>::image_snapshot_create(I *ictx, uint32_t flags,
+  if (r == -EOPNOTSUPP) {
+    ldout(cct, 5) << "group mirroring not supported by OSD" << dendl;
+    return r;
+  } else if (r < 0 && r != -ENOENT) {
+    lderr(cct) << "failed to retrieve mirror group metadata: "
+               << cpp_strerror(r) << dendl;
+    return r;
+  }
+
+  auto mode = static_cast<rbd_mirror_image_mode_t>(
+      mirror_group.mirror_image_mode);
+  if (mode != mirror_image_mode) {
+    lderr(cct) << "mirroring for group " << group_name
+               << " already enabled with different mirror image mode " << mode
+               << dendl;
+    return -EINVAL;
+  }
+
+  if (mirror_group.state == cls::rbd::MIRROR_GROUP_STATE_ENABLED) {
+    ldout(cct, 10) << "mirroring for group " << group_name
+                   << " already enabled" << dendl;
+    return 0;
+  }
+
+  if (mirror_group.state == cls::rbd::MIRROR_GROUP_STATE_ENABLING) {
+    lderr(cct) << "enabling mirroring for group " << group_name
+               << " either in progress or was interrupted" << dendl;
+    return -EINVAL;
+  }
+
+  std::string group_snap_id = librbd::util::generate_uuid(group_ioctx);
+  cls::rbd::GroupSnapshot group_snap{
+      group_snap_id,
+      cls::rbd::MirrorGroupSnapshotNamespace{},
+      calc_ind_mirror_snap_name(group_ioctx.get_id(), group_id, group_snap_id),
+      cls::rbd::GROUP_SNAPSHOT_STATE_INCOMPLETE};
+
+  std::vector<uint64_t> quiesce_requests;
+  std::vector<I *> image_ctxs;
+  r = prepare_group_images(group_ioctx, group_id, &image_ctxs,
+                           &group_snap, quiesce_requests,
+                           cls::rbd::MIRROR_SNAPSHOT_STATE_PRIMARY,
+                           flags);
+  if (r != 0) {
+    return r;
+  }
+
+  auto image_count = image_ctxs.size();
+  std::string group_header_oid = librbd::util::group_header_name(group_id);
+  std::vector<uint64_t> snap_ids(image_ctxs.size(), CEPH_NOSNAP);
+  int ret_code = is_images_mirror_enabled(&image_ctxs);
+  if (!ret_code) {
+    uuid_d uuid_gen;
+    uuid_gen.generate_random();
+    mirror_group = {uuid_gen.to_string(),
+      static_cast<cls::rbd::MirrorImageMode>(mirror_image_mode),
+      cls::rbd::MIRROR_GROUP_STATE_ENABLING};
+    r = cls_client::mirror_group_set(&group_ioctx, group_id, mirror_group);
+    if (r < 0) {
+      lderr(cct) << "failed to set mirroring group metadata: "
+                 << cpp_strerror(r) << dendl;
+      ret_code = r;
+    }
+  }
+
+  if (ret_code) {
+    remove_interim_snapshots(group_ioctx, group_header_oid,
+                             &image_ctxs, &group_snap);
+    goto cleanup;
+  }
+
+  for (size_t i = 0; i < image_ctxs.size(); i++) {
+    r = image_enable(image_ctxs[i], group_ioctx.get_id(), group_id,
+                     group_snap_id, mirror_image_mode, false, &snap_ids[i]);
+    if (r < 0) {
+      if (ret_code == 0) {
+        ret_code = r;
+        break;
+      }
+    }
+    group_snap.snaps[i].snap_id = snap_ids[i];
+  }
+
+  if (ret_code) {
```

Comment:
Would it be a good idea to release the locks immediately so as not to hold up client ops?